website/blog/2022-07-10-presto-parquet-column-encryption.md

```diff
@@ -0,0 +1,43 @@
+
+Currently,  Parquet™ supports 2 encryption algorithms: AES-GCM and AES-CTR. AES-GCM is an authenticated encryption algorithm that can prevent unauthenticated writing. Besides the data confidentiality (encryption), it supports 2 levels of integrity verification/authentication: of the data (default), and of the data combined with an optional additional authenticated data (AAD), which is a free text to be signed, together with the data. However, the AAD is required to be stored separately, in a KV store as an example, from the file itself, while the AAD metadata/index is saved in the Parquet™ file itself. The decryption application reads the AAD metadata/index first from the Parquet™ file and then reads AAD from the KV store before it can decrypt the Parquet™-encrypted data. More details about AAD can be found in RFC Using AES-CCM and AES-GCM Authenticated Encryption(https://tools.ietf.org/html/rfc5084).
+AES-CTR doesn’t require integrity verification/authentication. As a result, it is faster than AES-GCM. The benchmarking result shows AES-CTR is 3 times faster than AES-GCM in single-thread applications in Java 9, or 4.5 times faster than AES-GCM in single-thread applications in Java 8.
+
+## Parquet Column Encryption in Presto
+There are two Parquet writing code paths in current Presto repository native writer and legacy writer. The legacy writer relies on parquet-mr repository, which already has the implementation of Parquet modular encryption. As a result, Presto legacy writer already supports Parquet encryption in the newest release. The work to support native writer Parquet encryption is in progress.
+
+Parquet reader was rewritten in Presto, which requires the decryption code in parquet-mr need to be ported in Presto.  The code change [PR-17791](https://github.com/prestodb/presto/pull/17791) was merged into Presto repository and it will be released in next release Presto-0.274.
+
+One example about the Parquet encryption and decryption in Presto is described in the diagram below. In this example, there are only 2 columns (c1, c2). C2 is defined as a sensitive column while c1 is not. After the Parquet™ encryption, c2 is encrypted before being sent to storage, which could be HDFS or cloud storage like S3, GCS, Azure Blob, etc. On the reading path, the crypto metadata is stored in each file ([format](https://github.com/apache/parquet-format/blob/master/Encryption.md)) and the Parquet™ library uses that to determine what key to be retrieved to decrypt the data. The KMS client is wrapped up with the same plugin.
+
+![Presto Parquet Column Encryption Example](/img/blog/2022-07-10-presto-parquet-column-encryption/presto_paraquet_encryption_example.png)
```

Comment:
hmmm I don't see this binary being attached with this blog....

Reply:
Sorry, just added it now. 

> Would be good to paste a screenshot of the rendered webpage so we can review it more easily....

Where to paste the screen shot? Or you can just open it in intellilj and it will display